src/IO/H5/ReadTabulated.cpp

```diff
@@ -0,0 +1,108 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "IO/H5/ReadTabulated.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <hdf5.h>
+
+#include "DataStructures/Matrix.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "IO/H5/CheckH5.hpp"
+#include "IO/H5/OpenGroup.hpp"
+#include "IO/H5/Type.hpp"
+
+namespace {
+/*!
+ * Given a vector of contiguous data and an array giving the dimensions of a
+ * rank-3 tensor returns a `vector<Matrix>` representing the tensor.
+ */
+std::vector<Matrix> vector_to_rank3_tensor(
+    const std::vector<double>& raw_data,
+    const std::array<hsize_t, 3>& size) noexcept {
+  std::vector<Matrix> temp(size[0], Matrix(size[1], size[2]));
+  for (size_t i = 0; i < size[0]; ++i) {
+    for (size_t j = 0; j < size[1]; ++j) {
+      for (size_t k = 0; k < size[2]; ++k) {
+        temp[i](j, k) = raw_data[k + j * size[2] + i * size[2] * size[1]];
+      }
+    }
+  }
+  return temp;
+}
+}  // namespace
+
+namespace h5 {
+Tabulated::Tabulated(bool exists, detail::OpenGroup&& group,
+                     const hid_t location, const std::string& /*name*/)
+    : exists_(exists), group_(std::move(group)), location_(location) {}
+
+std::vector<double> Tabulated::get_rank1_dataset(
+    const std::string& dataset_name) const {
+  hid_t dataset_id{-1};
```

Comment:
DRY: Don't Repeat Yourself

Can you factor out the opening code into a private member function that is called from both the rank1 and rank3 one? I mean to the end of the `else` (line 53)